src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs

```diff
@@ -285,23 +285,83 @@ public DependencyContext Build(string[] userRuntimeAssemblies = null)
 
             foreach (var directReference in directAndDependencyReferences)
             {
-                var runtimeLibrary = new RuntimeLibrary(
+                var runtimeLibrary = new ModifiableRuntimeLibrary(new RuntimeLibrary(
                     type: "reference",
                     name: GetReferenceLibraryName(directReference),
                     version: directReference.Version,
                     hash: string.Empty,
-                    runtimeAssemblyGroups: new[] { new RuntimeAssetGroup(string.Empty, new[] { CreateRuntimeFile(directReference.FileName, directReference.FullPath) }) },
-                    nativeLibraryGroups: new RuntimeAssetGroup[] { },
+                    runtimeAssemblyGroups: [new RuntimeAssetGroup(string.Empty, [CreateRuntimeFile(directReference.FileName, directReference.FullPath)])],
+                    nativeLibraryGroups: [],
                     resourceAssemblies: CreateResourceAssemblies(directReference.ResourceAssemblies),
-                    dependencies: Enumerable.Empty<Dependency>(),
+                    dependencies: [],
                     path: null,
                     hashPath: null,
                     runtimeStoreManifestName: null,
-                    serviceable: false);
+                    serviceable: false));
 
                 runtimeLibraries.Add(runtimeLibrary);
             }
 
+            /*
+             * We now need to modify runtimeLibraries to eliminate those that don't have any runtime assets. We follow the following steps:
+             * 0. Construct a reverse dependencies list: all runtimeLibraries that depend on this one
+             * 1. If runtimeAssemblyGroups, nativeLibraryGroups, dependencies, and resourceAssemblies are all empty, remove this runtimeLibrary as well as any dependencies on it.
+             * 2. Add all runtimeLibraries to a list of to-be-processed libraries called libraryCandidatesForRemoval
+             * 3. libraryCandidatesForRemoval.Pop() --> if there are no runtimeAssemblyGroups, nativeLibraryGroups, or resourceAssemblies, and either dependencies is empty or all
+             *      dependencies have something else that depends on them, remove it (and from libraryCandidatesForRemoval), adding everything that depends on this to 
+             *      libraryCandidatesForRemoval if it isn't already there
+             * Repeat 3 until libraryCandidatesForRemoval is empty
+             */
+            var references = runtimeLibraries.ToDictionary(lib => lib.Library.Name, lib => lib);
+            foreach (var reference in runtimeLibraries)
```

Comment:
"Reference" seems to mean something slightly different than "library".  Is there a reason not to name these `libraries` and `library` instead of `references` and `reference`?